File: javascript/src/theme.ts
import { TestStepResultStatus } from '@cucumber/messages'

import { Style, Theme } from './types'

const CUCUMBER_STATUS_COLORS: Record<TestStepResultStatus, Style> = {
  [TestStepResultStatus.AMBIGUOUS]: 'magenta',
  [TestStepResultStatus.FAILED]: 'red',
  [TestStepResultStatus.PASSED]: 'green',
  [TestStepResultStatus.PENDING]: 'cyan',
  [TestStepResultStatus.SKIPPED]: 'yellow',
  [TestStepResultStatus.UNDEFINED]: 'blue',
  [TestStepResultStatus.UNKNOWN]: 'gray',
}

const CUCUMBER_STATUS_ICONS: Record<TestStepResultStatus, string> = {
  [TestStepResultStatus.AMBIGUOUS]: '✘',
  [TestStepResultStatus.FAILED]: '✘',
  [TestStepResultStatus.PASSED]: '✔',
  [TestStepResultStatus.PENDING]: '■',
  [TestStepResultStatus.SKIPPED]: '↷',
  [TestStepResultStatus.UNDEFINED]: '■',
  [TestStepResultStatus.UNKNOWN]: ' ',
}

const CUCUMBER_PROGRESS_ICONS: Record<TestStepResultStatus, string> = {
  [TestStepResultStatus.AMBIGUOUS]: 'A',
  [TestStepResultStatus.FAILED]: 'F',
  [TestStepResultStatus.PASSED]: '.',
  [TestStepResultStatus.PENDING]: 'P',
  [TestStepResultStatus.SKIPPED]: '-',
  [TestStepResultStatus.UNDEFINED]: 'U',
  [TestStepResultStatus.UNKNOWN]: '?',
}

export const CUCUMBER_THEME: Theme = {
  attachment: 'blue',
  feature: {
    keyword: 'bold',
  },
  location: 'blackBright',
  status: {
    all: CUCUMBER_STATUS_COLORS,
    icon: CUCUMBER_STATUS_ICONS,
    progress: CUCUMBER_PROGRESS_ICONS,
  },
  rule: {
    keyword: 'bold',
  },
  scenario: {
    keyword: 'bold',
  },
  step: {
    argument: 'bold',
    keyword: 'bold',
  },
  symbol: {
    bullet: '•',
  },
}

export const DEMO_THEME: Theme = {
  attachment: 'blue',
  dataTable: {
    all: 'blackBright',
    border: 'dim',
    content: 'italic',
  },
  docString: {
    all: 'blackBright',
    content: 'italic',
    delimiter: 'dim',
    mediaType: 'bold',
  },
  feature: {
    all: 'bgBlue',
    keyword: 'bold',
    name: 'italic',
  },
  location: 'blackBright',
  status: {
    all: CUCUMBER_STATUS_COLORS,
  },
  rule: {
    all: 'bgBlue',
    keyword: 'bold',
    name: 'italic',
  },
  scenario: {
    all: 'bgBlue',
    keyword: 'bold',
    name: 'italic',
  },
  step: {
    argument: 'bold',
    keyword: 'bold',
    text: 'italic',
  },
  tag: ['yellow', 'bold'],
  symbol: {
    bullet: '•',
  },
}

export const PLAIN_THEME: Theme = {
  status: {
    icon: CUCUMBER_STATUS_ICONS,
    progress: CUCUMBER_PROGRESS_ICONS,
  },
  symbol: {
    bullet: '-',
  },
}

export const NONE_THEME: Theme = {}

File: ruby/lib/cucumber/pretty_formatter/theme.rb
# frozen_string_literal: true

module Cucumber
  module PrettyFormatter
    module Styles
      CODES = {
        black_bright: [90, 39],
        blue: [34, 39],
        bold: [1, 22],
        bg_blue: [44, 49],
        cyan: [36, 39],
        dim: [2, 22],
        green: [32, 39],
        italic: [3, 23],
        magenta: [35, 39],
        red: [31, 39],
        yellow: [33, 39]
      }.freeze
    end

    CUCUMBER_STATUS_COLORS = {
      'AMBIGUOUS' => :magenta,
      'FAILED' => :red,
      'PASSED' => :green,
      'PENDING' => :cyan,
      'SKIPPED' => :yellow,
      'UNDEFINED' => :blue,
      'UNKNOWN' => :black_bright
    }.freeze

    CUCUMBER_STATUS_ICONS = {
      'AMBIGUOUS' => '✘',
      'FAILED' => '✘',
      'PASSED' => '✔',
      'PENDING' => '■',
      'SKIPPED' => '↷',
      'UNDEFINED' => '■',
      'UNKNOWN' => ' '
    }.freeze

    CUCUMBER_PROGRESS_ICONS = {
      'AMBIGUOUS' => 'A',
      'FAILED' => 'F',
      'PASSED' => '.',
      'PENDING' => 'P',
      'SKIPPED' => '-',
      'UNDEFINED' => 'U',
      'UNKNOWN' => '?'
    }.freeze

    CUCUMBER_THEME = {
      affix: :italic,
      attachment: :blue,
      feature: { keyword: :bold },
      location: :black_bright,
      status: {
        all: CUCUMBER_STATUS_COLORS,
        icon: CUCUMBER_STATUS_ICONS,
        progress: CUCUMBER_PROGRESS_ICONS
      },
      rule: { keyword: :bold },
      scenario: { keyword: :bold },
      step: { argument: :bold, keyword: :bold },
      symbol: { bullet: '•' }
    }.freeze

    DEMO_THEME = {
      affix: :italic,
      attachment: :blue,
      data_table: { all: :black_bright, border: :dim, content: :italic },
      doc_string: { all: :black_bright, content: :italic, delimiter: :dim, media_type: :bold },
      feature: { all: :bg_blue, keyword: :bold, name: :italic },
      location: :black_bright,
      status: { all: CUCUMBER_STATUS_COLORS },
      rule: { all: :bg_blue, keyword: :bold, name: :italic },
      scenario: { all: :bg_blue, keyword: :bold, name: :italic },
      step: { argument: :bold, keyword: :bold, text: :italic },
      tag: %i[yellow bold],
      symbol: { bullet: '•' }
    }.freeze

    NONE_THEME = {}.freeze

    PLAIN_THEME = {
      status: {
        icon: CUCUMBER_STATUS_ICONS,
        progress: CUCUMBER_PROGRESS_ICONS
      },
      symbol: { bullet: '-' }
    }.freeze
  end
end
